packages/react-reconciler/src/ReactFiberHooks.new.js

```diff
@@ -1947,17 +1946,21 @@ function mountEvent<Args, Return, F: (...Array<Args>) => Return>(
     return eventFn._impl.apply(undefined, arguments);
   };
   eventFn._impl = callback;
+  return eventFn;
+}
 
+function mountEvent<Args, Return, F: (...Array<Args>) => Return>(
+  callback: F,
+): EventFunctionWrapper<Args, Return, F> {
+  const eventFn = wrapEventFunction(callback);
   useEventImpl(eventFn, callback);
-  hook.memoizedState = eventFn;
   return eventFn;
 }
 
 function updateEvent<Args, Return, F: (...Array<Args>) => Return>(
   callback: F,
 ): EventFunctionWrapper<Args, Return, F> {
-  const hook = updateWorkInProgressHook();
-  const eventFn = hook.memoizedState;
+  const eventFn = wrapEventFunction(callback);
```

Comment:
This won't work because `useEventImpl` will only mutate the most recent "event wrapper" during the commit phase. All the previously returned events will point to their original implementation, which is not what we want. They should be updated, too.

Instead of putting the function wrapper in the effect queue, we should go back to placing a shared mutable object in the queue. That shared mutable object is what we'll store on `hook.memoizedState`. Then update that in the commit phase.

Reply:
I would expect this to break existing tests though so maybe there's a reason I'm missing why this actually does work. I'm not seeing it, though.

Reply:
I think it's because none of the existing tests happen to cover this case. For example, in this test, even though the event isn't part of the deps list, the URL is, so the effect happens to rerun and get the latest implementation, anyway: https://github.com/facebook/react/blob/60326eff8a7034376fabce8e692c01cb05d5a06f/packages/react-reconciler/src/__tests__/useEvent-test.js#L701-L709